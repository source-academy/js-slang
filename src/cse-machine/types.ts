import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
