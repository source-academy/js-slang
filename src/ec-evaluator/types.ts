import * as es from 'estree'

import { Context } from '..'
import { Environment } from '../types'
import { Agenda, Stash } from './interpreter'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED_IF_NEEDED = 'PushUndefinedIfNeeded',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
  srcNode: es.Node
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
  srcNode: es.Node
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
  srcNode: es.Node
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
  srcNode: es.Node
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
  srcNode: es.Node
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

export type CmdEvaluator = (
  command: AgendaItem,
  context: Context,
  agenda: Agenda,
  stash: Stash
) => void

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
