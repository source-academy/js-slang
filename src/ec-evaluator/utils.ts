import * as es from 'estree'
import { uniqueId } from 'lodash'

import { Context } from '..'
import * as errors from '../errors/errors'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import Closure from '../interpreter/closure'
import { Environment, Frame, Value } from '../types'
import * as ast from '../utils/astCreator'
import * as instr from './instrCreator'
import { Agenda } from './interpreter'
import { AgendaItem, AssmtInstr, Instr, InstrType } from './types'

/**
 * Stack is implemented for agenda and stash registers.
 */
interface IStack<T> {
  push(...items: T[]): void
  pop(): T | undefined
  peek(): T | undefined
  size(): number
}

export class Stack<T> implements IStack<T> {
  // Bottom of the array is at index 0
  private storage: T[] = []

  public constructor() {}

  public push(...items: T[]): void {
    for (const item of items) {
      this.storage.push(item)
    }
  }

  public pop(): T | undefined {
    return this.storage.pop()
  }

  public peek(): T | undefined {
    if (this.size() === 0) {
      return undefined
    }
    return this.storage[this.size() - 1]
  }

  public size(): number {
    return this.storage.length
  }
}

/**
 * Typeguard for Instr to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an instruction and false otherwise.
 */
export const isInstr = (command: AgendaItem): command is Instr => {
  return (command as Instr).instrType !== undefined
}

/**
 * Typeguard for esNode to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an esNode and false if it is an instruction.
 */
export const isNode = (command: AgendaItem): command is es.Node => {
  return (command as es.Node).type !== undefined
}

/**
 * Typeguard for esIdentifier. To verify if an esNode is an esIdentifier.
 *
 * @param node an esNode
 * @returns true if node is an esIdentifier, false otherwise.
 */
export const isIdentifier = (node: es.Node): node is es.Identifier => {
  return (node as es.Identifier).name !== undefined
}

/**
 * Typeguard for esBlockStatement. To verify if a function body is a block statement.
 *
 * @param body the function body
 * @returns true if node is an esIdentifier, false otherwise.
 */
export const isExpressionBody = (
  body: es.BlockStatement | es.Expression
): body is es.Expression => {
  return body.type !== 'BlockStatement'
}

/**
 * Typeguard for AssmtInstr. To verify if an instruction is an assignment instruction.
 *
 * @param instr an instruction
 * @returns true if instr is an AssmtInstr, false otherwise.
 */
export const isAssmtInstr = (instr: Instr): instr is AssmtInstr => {
  return instr.instrType === InstrType.ASSIGNMENT
}

/**
 * A helper function for handling sequences of statements.
 * Statements must be pushed in reverse order, and each statement is separated by a pop
 * instruction so that only the result of the last statement remains on stash.
 * Value producing statements have an extra pop instruction.
 *
 * @param seq Array of statements.
 * @returns Array of commands to be pushed into agenda.
 */
export const handleSequence = (seq: es.Statement[]): AgendaItem[] => {
  const result: AgendaItem[] = []
  let valueProduced = false
  for (const command of seq) {
    if (valueProducing(command)) {
      valueProduced ? result.push(instr.popInstr()) : (valueProduced = true)
    }
    result.push(command)
  }
  return result.reverse()
}

/**
 * This function is used for ConditionalExpressions and IfStatements, to create the sequence
 * of agenda items to be added.
 */
export const reduceConditional = (
  node: es.IfStatement | es.ConditionalExpression
): AgendaItem[] => {
  return [instr.branchInstr(node.consequent, node.alternate, node), node.test]
}

/**
 * To determine if an agenda item is value producing. JavaScript distinguishes valu producing
 * statements and non-value producing statements.
 * Refer to https://sourceacademy.nus.edu.sg/sicpjs/4.1.2 exercise 4.8.
 *
 * @param command Agenda item to determine if it is value producing.
 * @returns true if it is value producing, false otherwise.
 */
export const valueProducing = (command: es.Node): boolean => {
  const type = command.type
  return (
    type !== 'VariableDeclaration' &&
    type !== 'FunctionDeclaration' &&
    type !== 'ContinueStatement' &&
    type !== 'BreakStatement' &&
    type !== 'ReturnStatement' &&
    (type !== 'BlockStatement' || command.body.some(valueProducing))
  )
}

/**
 * Environments
 */

export const currentEnvironment = (context: Context) => context.runtime.environments[0]

export const createEnvironment = (
  closure: Closure,
  args: Value[],
  callExpression: es.CallExpression
): Environment => {
  const environment: Environment = {
    name: isIdentifier(callExpression.callee) ? callExpression.callee.name : closure.functionName,
    tail: closure.environment,
    head: {},
    id: uniqueId(),
    callExpression: {
      ...callExpression,
      arguments: args.map(ast.primitive)
    }
  }
  closure.node.params.forEach((param, index) => {
    environment.head[(param as es.Identifier).name] = args[index]
  })
  return environment
}

export const popEnvironment = (context: Context) => context.runtime.environments.shift()

export const pushEnvironment = (context: Context, environment: Environment) => {
  context.runtime.environments.unshift(environment)
  context.runtime.environmentTree.insert(environment)
}

export const createBlockEnvironment = (
  context: Context,
  name = 'blockEnvironment',
  head: Frame = {}
): Environment => {
  return {
    name,
    tail: currentEnvironment(context),
    head,
    id: uniqueId()
  }
}

/**
 * Variables
 */

const DECLARED_BUT_NOT_YET_ASSIGNED = Symbol('Used to implement hoisting')

export function declareIdentifier(
  context: Context,
  name: string,
  node: es.Node,
  environment: Environment
) {
  if (environment.head.hasOwnProperty(name)) {
    const descriptors = Object.getOwnPropertyDescriptors(environment.head)

    return handleRuntimeError(
      context,
      new errors.VariableRedeclaration(node, name, descriptors[name].writable)
    )
  }
  environment.head[name] = DECLARED_BUT_NOT_YET_ASSIGNED
  return environment
}

function declareVariables(
  context: Context,
  node: es.VariableDeclaration,
  environment: Environment
) {
  for (const declaration of node.declarations) {
    declareIdentifier(context, (declaration.id as es.Identifier).name, node, environment)
  }
}

export function declareFunctionsAndVariables(
  context: Context,
  node: es.BlockStatement,
  environment: Environment
) {
  let hasDeclarations: boolean = false
  for (const statement of node.body) {
    switch (statement.type) {
      case 'VariableDeclaration':
        declareVariables(context, statement, environment)
        hasDeclarations = true
        break
      case 'FunctionDeclaration':
        declareIdentifier(context, (statement.id as es.Identifier).name, statement, environment)
        hasDeclarations = true
        break
    }
  }
  return hasDeclarations
}

export function defineVariable(
  context: Context,
  name: string,
  value: Value,
  constant = false,
  node: es.VariableDeclaration | es.ImportDeclaration
) {
  const environment = currentEnvironment(context)

  if (environment.head[name] !== DECLARED_BUT_NOT_YET_ASSIGNED) {
    return handleRuntimeError(context, new errors.VariableRedeclaration(node, name, !constant))
  }

  Object.defineProperty(environment.head, name, {
    value,
    writable: !constant,
    enumerable: true
  })

  return environment
}

export const getVariable = (context: Context, name: string, node: es.Identifier) => {
  let environment: Environment | null = currentEnvironment(context)
  while (environment) {
    if (environment.head.hasOwnProperty(name)) {
      if (environment.head[name] === DECLARED_BUT_NOT_YET_ASSIGNED) {
        return handleRuntimeError(context, new errors.UnassignedVariable(name, node))
      } else {
        return environment.head[name]
      }
    } else {
      environment = environment.tail
    }
  }
  return handleRuntimeError(context, new errors.UndefinedVariable(name, node))
}

export const setVariable = (
  context: Context,
  name: string,
  value: any,
  node: es.AssignmentExpression
) => {
  let environment: Environment | null = currentEnvironment(context)
  while (environment) {
    if (environment.head.hasOwnProperty(name)) {
      if (environment.head[name] === DECLARED_BUT_NOT_YET_ASSIGNED) {
        break
      }
      const descriptors = Object.getOwnPropertyDescriptors(environment.head)
      if (descriptors[name].writable) {
        environment.head[name] = value
        return undefined
      }
      return handleRuntimeError(context, new errors.ConstAssignment(node, name))
    } else {
      environment = environment.tail
    }
  }
  return handleRuntimeError(context, new errors.UndefinedVariable(name, node))
}

export const handleRuntimeError = (context: Context, error: RuntimeSourceError) => {
  context.errors.push(error)
  throw error
}

export const checkNumberOfArguments = (
  context: Context,
  callee: Closure | Value,
  args: Value[],
  exp: es.CallExpression
) => {
  if (callee instanceof Closure) {
    // User-defined or Pre-defined functions
    const params = callee.node.params
    const hasVarArgs = params[params.length - 1]?.type === 'RestElement'
    if (hasVarArgs ? params.length - 1 > args.length : params.length !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(
          exp,
          hasVarArgs ? params.length - 1 : params.length,
          args.length,
          hasVarArgs
        )
      )
    }
  } else {
    // Pre-built functions
    const hasVarArgs = callee.minArgsNeeded != undefined
    if (hasVarArgs ? callee.minArgsNeeded > args.length : callee.length !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(
          exp,
          hasVarArgs ? callee.minArgsNeeded : callee.length,
          args.length,
          hasVarArgs
        )
      )
    }
  }
  return undefined
}

/**
 * This function can be used to check for a stack overflow.
 * The current limit is set to be an agenda size of 1.0 x 10^5, if the agenda
 * flows beyond this limit an error is thrown.
 * This corresponds to about 10mb of space according to tests ran.
 */
export const checkStackOverFlow = (context: Context, agenda: Agenda) => {
  if (agenda.size() > 100000) {
    const stacks: es.CallExpression[] = []
    let counter = 0
    for (
      let i = 0;
      counter < errors.MaximumStackLimitExceeded.MAX_CALLS_TO_SHOW &&
      i < context.runtime.environments.length;
      i++
    ) {
      if (context.runtime.environments[i].callExpression) {
        stacks.unshift(context.runtime.environments[i].callExpression!)
        counter++
      }
    }
    handleRuntimeError(
      context,
      new errors.MaximumStackLimitExceeded(context.runtime.nodes[0], stacks)
    )
  }
}
