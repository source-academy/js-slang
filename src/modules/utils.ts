/**
 * For tabs that are export default declarations, we need to remove
 * the `export default` bit from the front before they can be loaded
 * by js-slang
 */
export function evalRawTab(text: string) {
  if (text.startsWith('export default')) {
    text = text.substring(14)
  }

  return eval(text)
}
