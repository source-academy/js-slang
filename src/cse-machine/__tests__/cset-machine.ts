import { Chapter, Variant } from '../../types'
import { expectParsedError } from '../../utils/testing'

// CSET tests for Scheme (mostly error testing for
// the runtime verification of scheme syntax forms.)
const optionECScm = { chapter: Chapter.FULL_SCHEME, variant: Variant.EXPLICIT_CONTROL }

test('evaluating null throws error', () => {
  return expectParsedError(
    `
    (eval '())
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Cannot evaluate null"`)
})

test('evaluating a poorly formed lambda throws error', () => {
  return expectParsedError(
    `
    (eval '(lambda (1 2 3) x))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Invalid arguments for lambda!"`)
})

test('evaluating a poorly formed define throws error (insufficient arguments)', () => {
  return expectParsedError(
    `
    (eval '(define))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define requires at least 2 arguments!"`)
})

test('evaluating a poorly formed define throws error (too many arguments)', () => {
  return expectParsedError(
    `
    (eval '(define x 1 2 3))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define requires 2 arguments!"`)
})

test('evaluating a poorly formed define throws error (ill formed define-function)', () => {
  return expectParsedError(
    `
    (eval '(define (x 1 2 3) 4))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Invalid arguments for lambda!"`)
})

test('evaluating a poorly formed define throws error (attempt to redefine special form)', () => {
  return expectParsedError(
    `
    (eval '(define (if x y) 4))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Cannot shadow special form if with a definition!"`)
})

test('evaluating a poorly formed set! throws error (insufficient arguments)', () => {
  return expectParsedError(
    `
    (eval '(set!))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: set! requires 2 arguments!"`)
})

test('evaluating a poorly formed set! throws error (too many arguments)', () => {
  return expectParsedError(
    `
    (eval '(set! x 1 2 3))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: set! requires 2 arguments!"`)
})

test('evaluating a poorly formed set! throws error (attempt to set! special form)', () => {
  return expectParsedError(
    `
    (eval '(set! if 4))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Cannot overwrite special form if with a value!"`)
})

test('evaluating a poorly formed if throws error (insufficient arguments)', () => {
  return expectParsedError(
    `
    (eval '(if))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: if requires at least 2 arguments!"`)
})

test('evaluating a poorly formed if throws error (too many arguments)', () => {
  return expectParsedError(
    `
    (eval '(if #t 1 2 3))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: if requires at most 3 arguments!"`)
})

test('evaluating a poorly formed begin throws error (insufficient arguments)', () => {
  return expectParsedError(
    `
    (eval '(begin))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: begin requires at least 1 argument!"`)
})

test('evaluating a poorly formed quote throws error (insufficient arguments)', () => {
  return expectParsedError(
    `
    (eval '(quote))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: quote requires 1 argument!"`)
})

test('evaluating a poorly formed quote throws error (too many arguments)', () => {
  return expectParsedError(
    `
    (eval '(quote x y))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: quote requires 1 argument!"`)
})

test('evaluating a poorly formed define-syntax throws error (insufficient arguments)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define-syntax requires 2 arguments!"`)
})

test('evaluating a poorly formed define-syntax throws error (too many arguments)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x 1 2 3))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define-syntax requires 2 arguments!"`)
})

test('evaluating a poorly formed define-syntax throws error (syntax is not a symbol)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax 1 4))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define-syntax requires a symbol!"`)
})

test('evaluating a poorly formed define-syntax throws error (attempt to shadow special form)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax if 4))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Cannot shadow special form if with a macro!"`)
})

test('evaluating a poorly formed define-syntax throws error (no syntax-rules list)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x 1))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define-syntax requires a syntax-rules transformer!"`)
})

test('evaluating a poorly formed define-syntax throws error (list is not syntax-rules)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x (foo bar)))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: define-syntax requires a syntax-rules transformer!"`)
})

test('evaluating a poorly formed define-syntax throws error (syntax-rules too few arguments)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x (syntax-rules)))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: syntax-rules requires at least 2 arguments!"`)
})

test('evaluating a poorly formed define-syntax throws error (syntax-rules has poor literals list)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x (syntax-rules x (1 1))))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Invalid syntax-rules literals!"`)
})

test('evaluating a poorly formed define-syntax throws error (syntax-rules has non-symbol literals)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x (syntax-rules (1 2) (1 1))))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Invalid syntax-rules literals!"`)
})

test('evaluating a poorly formed define-syntax throws error (syntax-rules has non-list rules)', () => {
  return expectParsedError(
    `
    (eval '(define-syntax x (syntax-rules (x) 1)))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: Invalid syntax-rules rule!"`)
})

test('evaluating a syntax-rules expression (should not exist outside of define-syntax)', () => {
  return expectParsedError(
    `
    (eval '(syntax-rules (x) (1 1)))
    `,
    optionECScm
  ).toMatchInlineSnapshot(`"Error: syntax-rules must only exist within define-syntax!"`)
})
