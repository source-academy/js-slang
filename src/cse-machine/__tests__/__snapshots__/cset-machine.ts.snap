// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluating a poorly formed begin throws error (insufficient arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(begin))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: begin requires at least 1 argument!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define throws error (attempt to redefine special form): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define (if x y) 4))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Cannot shadow special form if with a definition!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define throws error (ill formed define-function): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define (x 1 2 3) 4))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Invalid arguments for lambda!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define throws error (insufficient arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define requires at least 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define throws error (too many arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define x 1 2 3))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define requires 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (attempt to shadow special form): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax if 4))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Cannot shadow special form if with a macro!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (insufficient arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define-syntax requires 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (list is not syntax-rules): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x (foo bar)))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define-syntax requires a syntax-rules transformer!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (no syntax-rules list): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x 1))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define-syntax requires a syntax-rules transformer!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (syntax is not a symbol): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax 1 4))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define-syntax requires a symbol!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (syntax-rules has non-list rules): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x (syntax-rules (x) 1)))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Invalid syntax-rules rule!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (syntax-rules has non-symbol literals): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x (syntax-rules (1 2) (1 1))))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Invalid syntax-rules literals!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (syntax-rules has poor literals list): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x (syntax-rules x (1 1))))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Invalid syntax-rules literals!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (syntax-rules too few arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x (syntax-rules)))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: syntax-rules requires at least 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed define-syntax throws error (too many arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(define-syntax x 1 2 3))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: define-syntax requires 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed if throws error (insufficient arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(if))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: if requires at least 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed if throws error (too many arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(if #t 1 2 3))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: if requires at most 3 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed lambda throws error: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(lambda (1 2 3) x))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Invalid arguments for lambda!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed quote throws error (insufficient arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(quote))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: quote requires 1 argument!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed quote throws error (too many arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(quote x y))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: quote requires 1 argument!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed set! throws error (attempt to set! special form): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(set! if 4))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Cannot overwrite special form if with a value!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed set! throws error (insufficient arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(set!))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: set! requires 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a poorly formed set! throws error (too many arguments): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(set! x 1 2 3))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: set! requires 2 arguments!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating a syntax-rules expression (should not exist outside of define-syntax): expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '(syntax-rules (x) (1 1)))
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: syntax-rules must only exist within define-syntax!",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`evaluating null throws error: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    (eval '())
    ",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Error: Cannot evaluate null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;
